health/manager.go

```diff
@@ -0,0 +1,303 @@
+	AddTargets(targets []*target.Target)
+}
+
+// NewManager creates a new Manager instance with internal target registry.
+// You should init configuration and writer before and pass it here
+func NewManager(ctx context.Context, config *Config, writer w.HealthWriter) Manager {
+	healthReg := newHealthRegistry()
+
+	return &healthManager{
+		registry: healthReg,
+		config:   config,
+		writer:   writer,
+	}
+}
+
+// healthManager implements Manager interface
+type healthManager struct {
+	registry healthRegistry
+	config   *Config
+	writer   w.HealthWriter
+}
+
+func (hm *healthManager) Start(ctx context.Context) {
+	logging.SetLogLevel(hm.config.LogLevel)
+	measurements := make(chan *targetMeasurement)
+	initCollector(hm.config.CollectionCycle, measurements)
+	go hm.listenMeasurements(ctx, measurements)
+
+	healthIn := make(chan *target.Health)
+	go hm.healthForwarder(ctx, healthIn)
+	go hm.writer.Start(healthIn)
+}
+
+func (hm *healthManager) AddTargets(targets []*target.Target) {
+	hm.registry.lock()
+	defer hm.registry.unlock()
+
+	for _, target := range targets {
+		hm.registry.setRawHealthForTarget(newRawHealth(target, target.MetricIDs))
+	}
+}
+
+// Health Manager listens for data from collector and stores it in a registry
+func (hm *healthManager) listenMeasurements(ctx context.Context, measurements <-chan *targetMeasurement) {
+	log := logging.GetLogger()
+	log.Info().Msg("Start to listen for collected measurements")
+	defer log.Info().Msg("Finish listening for measurements from collector")
+	for {
+		select {
+		case measurement := <-measurements:
+			err := hm.updateTargetHealthData(measurement)
+			if err != nil {
+				log.Error().AnErr("error", err).Msgf("Unable to update target with id %s", measurement.targetID)
+			}
+		case <-ctx.Done():
+			shutDownCollector()
+			return
+		}
+	}
+}
+
+func (hm *healthManager) updateTargetHealthData(measure *targetMeasurement) error {
+	var err error
+
+	hm.registry.lock()
+	defer hm.registry.unlock()
+
+	targetHealth, ok := hm.registry.getRawHealthForTarget(measure.targetID)
+	if !ok {
+		if !hm.config.RegistrationOnCollect {
+			return errTargetNotRegistered
+		}
+		targetHealth, err = hm.buildTargetFromMeasure(measure)
+		if err != nil {
+			return fmt.Errorf("Unable to register target automatically: %w", err)
+		}
+	}
+
+	switch measure.measureType {
+	case heartbeat:
+		targetHealth.heartBeat = true
+	case metric:
+		err := hm.updateTargetsMetric(targetHealth, measure)
+		if err != nil {
+			return fmt.Errorf("Unable to update target metric: %w", err)
+		}
+	case counterChange:
+		err := hm.updateTargetsCounter(targetHealth, measure)
+		if err != nil {
+			return fmt.Errorf("Unable to update target counter: %w", err)
+		}
+	case message:
+		if measure.message.AffectHealth {
+			targetHealth.healthy = false
+		}
+		targetHealth.messages = append(targetHealth.messages, measure.message)
+	case healthStatus:
+		targetHealth.healthy = measure.healthStatus
```

Comment:
need to reconcile with message based health

Reply:
Message is updated with HealthStatus field and can impact component health status in the same way as the appropriate separate measure